pylabrobot/heating_shaking/hamilton.py

```diff
@@ -0,0 +1,105 @@
+    Multiple Hamilton Heater Shakers can be connected to the same Heat Shaker Box. Each has A
+    separate 'shaker index'
+    """
+    assert shaker_index >= 0, "Shaker index must be non-negative"
+    self.shaker_index = shaker_index
+    self.command_id = 0
+    self.lock_initialized = False
+
+    HeaterShakerBackend.__init__(self)
+    USBBackend.__init__(self, id_vendor, id_product)
+
+  async def setup(self):
+    await USBBackend.setup(self)
+
+  async def stop(self):
+    await USBBackend.stop(self)
+
+  def _send_command(self, command: str, **kwargs):
+    assert len(command) == 2, "Command must be 2 characters long"
+    args = "".join([f"{key}{value}" for key, value in kwargs.items()])
+    USBBackend.write(
+      self,
+      f"T{self.shaker_index}{command}id{str(self.command_id).zfill(4)}{args}",
+    )
+
+    self.command_id = (self.command_id + 1) % 10_000
+    return USBBackend.read(self)
+
+  async def shake(
+    self,
+    speed: float = 800,
+    direction: Literal[0,1] = 0,
+    acceleration: int = 1_000,
+  ):
+    """
+    speed: steps per second
+    direction: 0 for positive, 1 for negative
+    acceleration: increments per second
+    """
+    int_speed = int(speed)
+    assert 20 <= int_speed <= 2_000, "Speed must be between 20 and 2_000"
+    assert direction in [0, 1], "Direction must be 0 or 1"
+    assert 500 <= acceleration <= 10_000, "Acceleration must be between 500 and 10_000"
+
+    if not self.lock_initialized:
+      await self._initialize_lock()
+    await self._move_plate_lock(PlateLockPosition.UNLOCKED)
+    await self._start_shaking(direction=direction, speed=int_speed, acceleration=acceleration)
+
+  async def stop_shaking(self):
+    """ Shaker `stop_shaking` implementation. """
+    await self._stop_shaking()
+    await self._wait_for_stop()
+    await self._move_plate_lock(PlateLockPosition.LOCKED)
+
+  async def _move_plate_lock(self, position: PlateLockPosition):
+    return self._send_command("LP", lp=position.value)
+
+  async def _initialize_lock(self):
+    """ Firmware command initialize lock. """
+    result = self._send_command("LI")
+    self.lock_initialized = True
+    return result
+
+  async def _start_shaking(self, direction: int, speed: int, acceleration: int):
+    """ Firmware command for starting shaking. """
+    speed_str = str(speed).zfill(4)
+    acceleration_str = str(acceleration).zfill(5)
+    return self._send_command("SB", st=direction, sv=speed_str, sr=acceleration_str)
+
+  async def _stop_shaking(self):
+    """ Firmware command for stopping shaking. """
+    return self._send_command("SC")
+
+  async def _wait_for_stop(self):
+    """ Firmware command for waiting for shaking to stop. """
+    return self._send_command("SW")
+
+  async def set_temperature(self, temperature: float):
+    raise NotImplementedError()
+
+  async def get_current_temperature(self) -> float:
+    raise NotImplementedError()
+
+  async def deactivate(self):
+    raise NotImplementedError()
```

Comment:
brought these back since the abstract super class demands it. Left them for the next adventurer who needs heat to implement 🙂